File: hackit/wireless/plugins/lua/scripts/mac_changer.lua
local iface = arg[1]
local action = arg[2] or "show"
local value = arg[3] or ""

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":"","ssid":"","channel":"","data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function get_current_mac()
  local f = io.popen("ip link show " .. iface .. " 2>/dev/null", "r")
  if not f then return nil end
  local out = f:read("*a")
  f:close()
  local mac = out:match("link/ether%s+([%x:]+)")
  return mac
end

local function get_original_mac()
  local f = io.popen("cat /sys/class/net/" .. iface .. "/addr_perm 2>/dev/null", "r")
  if f then
    local mac = f:read("*a")
    f:close()
    if mac and #mac > 0 then return mac:match("[%x:]+") end
  end
  local f2 = io.popen("ethtool -P " .. iface .. " 2>/dev/null | awk '{print $3}'", "r")
  if f2 then
    local mac = f2:read("*a")
    f2:close()
    if mac and #mac > 0 then return mac:match("[%x:]+") end
  end
  return get_current_mac()
end

local function random_mac()
  local r = io.popen("openssl rand -hex 6 2>/dev/null", "r")
  if not r then
    math.randomseed(os.time())
    local bytes = {}
    for i = 1, 6 do bytes[i] = string.format("%02x", math.random(0, 255)) end
    bytes[1] = string.format("%02x", (tonumber(bytes[1], 16) & 0xfe) | 0x02)
    return table.concat(bytes, ":")
  end
  local hex = r:read("*a")
  r:close()
  local mac = hex:gsub("(%x%x)(%x%x)(%x%x)(%x%x)(%x%x)(%x%x)", function(a, b, c, d, e, f)
    local fa = tonumber(a, 16) | 0x02
    fa = fa & 0xfe
    return string.format("%02x:%s:%s:%s:%s:%s", fa, b, c, d, e, f)
  end)
  return mac
end

local function do_change(new_mac)
  telemetry("change_start", '{"new_mac":' .. json_escape(new_mac) .. '}')
  os.execute("ip link set " .. iface .. " down 2>/dev/null")
  os.execute("ip link set " .. iface .. " address " .. new_mac .. " 2>/dev/null")
  os.execute("ip link set " .. iface .. " up 2>/dev/null")
  os.execute("sleep 1")
  local current = get_current_mac()
  if current and current == new_mac then
    telemetry("change_ok", '{"mac":' .. json_escape(current) .. '}')
    return true
  end
  telemetry("change_fail", '{"expected":' .. json_escape(new_mac) .. ',"got":' .. json_escape(current or "nil") .. '}')
  return false
end

local function restore_mac()
  local original = get_original_mac()
  if not original then
    telemetry("restore_fail", '{"error":"Cannot determine original MAC"}')
    return false
  end
  telemetry("restore_start", '{"original_mac":' .. json_escape(original) .. '}')
  os.execute("ip link set " .. iface .. " down 2>/dev/null")
  os.execute("ip link set " .. iface .. " address " .. original .. " 2>/dev/null")
  os.execute("ip link set " .. iface .. " up 2>/dev/null")
  os.execute("sleep 1")
  local current = get_current_mac()
  if current and current == original then
    telemetry("restore_ok", '{"mac":' .. json_escape(current) .. '}')
    return true
  end
  local cmd = "macchanger -p " .. iface .. " 2>/dev/null"
  os.execute(cmd)
  current = get_current_mac()
  telemetry("restore_result", '{"mac":' .. json_escape(current or "") .. '}')
  return true
end

local function show_mac()
  local current = get_current_mac()
  local original = get_original_mac()
  telemetry("mac_show", '{"current":' .. json_escape(current or "unknown") .. ',"original":' .. json_escape(original or "unknown") .. '}')
end

if not iface then
  telemetry("error", '{"error":"Usage: mac_changer.lua <interface> <action> [value]"}')
  os.exit(1)
end

local ok, err = pcall(function()
  if action == "show" then
    show_mac()
  elseif action == "random" then
    local mac = random_mac()
    do_change(mac)
  elseif action == "set" then
    if not value or #value == 0 then
      telemetry("error", '{"error":"No MAC address provided for set action"}')
      os.exit(1)
    end
    do_change(value)
  elseif action == "restore" then
    restore_mac()
  else
    telemetry("error", '{"error":"Unknown action: ' .. action .. '"}')
    os.exit(1)
  end
end)

if not ok then
  telemetry("error", '{"error":' .. json_escape(tostring(err)) .. '}')
  os.exit(1)
end

telemetry("complete", '{"action":' .. json_escape(action) .. '}')

File: hackit/wireless/plugins/ruby/scripts/mac_changer.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'open3'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

def get_current_mac(iface)
  stdout, _, status = Open3.capture3('ip', 'link', 'show', iface)
  return nil unless status.success?
  if stdout =~ /link\/ether\s+([a-fA-F0-9:]{17})/
    $1.downcase
  else
    nil
  end
end

def get_current_mac_ifconfig(iface)
  stdout, _, status = Open3.capture3('ifconfig', iface)
  return nil unless status.success?
  if stdout =~ /ether\s+([a-fA-F0-9:]{17})/
    $1.downcase
  else
    nil
  end
end

def set_mac_ip(iface, mac)
  cmds = [
    ['ip', 'link', 'set', iface, 'down'],
    ['ip', 'link', 'set', iface, 'address', mac],
    ['ip', 'link', 'set', iface, 'up']
  ]
  cmds.each do |cmd|
    stdout, stderr, status = Open3.capture3(*cmd)
    return false unless status.success?
  end
  true
end

def set_mac_macchanger(iface, mac)
  stdout, _, status = Open3.capture3('macchanger', '-m', mac, iface)
  status.success?
end

def set_mac_ifconfig(iface, mac)
  cmds = [
    ['ifconfig', iface, 'down'],
    ['ifconfig', iface, 'hw', 'ether', mac],
    ['ifconfig', iface, 'up']
  ]
  cmds.each do |cmd|
    _, _, status = Open3.capture3(*cmd)
    return false unless status.success?
  end
  true
end

def random_mac
  '%02x:%02x:%02x:%02x:%02x:%02x' % [
    rand(2..254) & 0xFE | 0x02,
    rand(256), rand(256), rand(256), rand(256), rand(256)
  ]
end

def restore_mac(iface)
  stdout, _, status = Open3.capture3('macchanger', '-p', iface)
  return status.success? if status.success?
  emit('mac_restore_fallback', iface, '', '', 0, { message: 'macchanger -p failed' })
  false
end

def show_mac_info(iface)
  current = get_current_mac(iface) || get_current_mac_ifconfig(iface)
  return { mac: current || 'unknown' }
end

begin
  iface = ARGV[0]
  action = ARGV[1] || 'show'
  value = ARGV[2]

  raise 'interface required' unless iface

  emit('mac_changer_start', iface, '', '', 0, { action: action })

  case action
  when 'show'
    info = show_mac_info(iface)
    emit('mac_show', iface, '', '', 0, info)

  when 'random'
    old_mac = get_current_mac(iface)
    new_mac = random_mac
    if set_mac_ip(iface, new_mac) || set_mac_macchanger(iface, new_mac) || set_mac_ifconfig(iface, new_mac)
      emit('mac_changed', iface, '', '', 0, { old_mac: old_mac, new_mac: new_mac, action: 'random' })
    else
      raise 'failed to set random MAC address'
    end

  when 'restore'
    old_mac = get_current_mac(iface)
    if restore_mac(iface)
      new_mac = get_current_mac(iface)
      emit('mac_restored', iface, '', '', 0, { old_mac: old_mac, new_mac: new_mac, action: 'restore' })
    else
      raise 'failed to restore MAC address'
    end

  when 'set'
    raise 'MAC address required for set action' unless value
    raise "invalid MAC format: #{value}" unless value =~ /^([a-fA-F0-9]{2}:){5}[a-fA-F0-9]{2}$/
    old_mac = get_current_mac(iface)
    if set_mac_ip(iface, value) || set_mac_macchanger(iface, value) || set_mac_ifconfig(iface, value)
      emit('mac_changed', iface, '', '', 0, { old_mac: old_mac, new_mac: value.downcase, action: 'set' })
    else
      raise 'failed to set MAC address'
    end
  else
    raise "unknown action: #{action}. Valid: show, random, restore, set"
  end

  emit('mac_changer_complete', iface, '', '', 0, { action: action })

rescue Interrupt
  emit('mac_changer_interrupted', iface || '', '', '', 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
end
